File: Source/AlphaTab/Util/Logger.cs
namespace AlphaTab.Util
{
    internal class Logger
    {
        public static LogLevel LogLevel { get; set; }

        static Logger()
        {
            LogLevel = LogLevel.Info;
        }

        public static void Debug(string category, string msg, object details = null)
        {
            Log(LogLevel.Debug, category, msg, details);
        }

        public static void Warning(string category, string msg, object details = null)
        {
            Log(LogLevel.Warning, category, msg, details);
        }

        public static void Info(string category, string msg, object details = null)
        {
            Log(LogLevel.Info, category, msg, details);
        }

        public static void Error(string category, string msg, object details = null)
        {
            Log(LogLevel.Error, category, msg, details);
        }

        public static void Log(LogLevel logLevel, string category, string msg, object details = null)
        {
            if (logLevel < LogLevel || LogLevel == LogLevel.None)
            {
                return;
            }

            Platform.Platform.Log(logLevel, category, msg, details);
        }
    }

    /// <summary>
    /// Defines all loglevels. 
    /// </summary>
    [JsonSerializable]
    public enum LogLevel
    {
        /// <summary>
        /// No logging
        /// </summary>
        None = 0,

        /// <summary>
        /// Debug level (internal details are displayed).
        /// </summary>
        Debug = 1,

        /// <summary>
        /// Info level (only important details are shown)
        /// </summary>
        Info = 2,

        /// <summary>
        /// Warning level
        /// </summary>
        Warning = 3,

        /// <summary>
        /// Error level.
        /// </summary>
        Error = 4
    }
}

File: src/util/Logger.ts
/**
 * Defines all loglevels.
 * @json
 */
export enum LogLevel {
    /**
     * No logging
     */
    None,
    /**
     * Debug level (internal details are displayed).
     */
    Debug,
    /**
     * Info level (only important details are shown)
     */
    Info,
    /**
     * Warning level
     */
    Warning,
    /**
     * Error level.
     */
    Error
}

export class Logger {
    public static logLevel: LogLevel = LogLevel.Info;

    private static format(category: string, msg: string): string {
        return `[AlphaTab][${category}] ${msg}`;
    }

    private static shouldLog(level: LogLevel): boolean {
        return Logger.logLevel !== LogLevel.None && level >= Logger.logLevel;
    }

    public static debug(category: string, msg: string, ...details: any[]): void {
        if (Logger.shouldLog(LogLevel.Debug)) {
            console.debug(Logger.format(category, msg), ...details);
        }
    }

    public static warning(category: string, msg: string, ...details: any[]): void {
        if (Logger.shouldLog(LogLevel.Warning)) {
            console.warn(Logger.format(category, msg), ...details);
        }
    }

    public static info(category: string, msg: string, ...details: any[]): void {
        if (Logger.shouldLog(LogLevel.Info)) {
            console.info(Logger.format(category, msg), ...details);
        }
    }

    public static error(category: string, msg: string, ...details: any[]): void {
        if (Logger.shouldLog(LogLevel.Error)) {
            console.error(Logger.format(category, msg), ...details);
        }
    }
}
